components/constellation/constellation.rs

```diff
@@ -1615,34 +1655,32 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
             .map(|pipeline_id| self.get_top_level_frame_for_pipeline(pipeline_id))
             .unwrap_or(self.root_frame_id);
 
-        let mut traversal_info = HashMap::new();
+        let mut size = 0;
+        let mut table = HashMap::new();
 
         match direction {
             TraversalDirection::Forward(delta) => {
-                let mut future = self.joint_session_future(top_level_frame_id);
-                for _ in 0..delta {
-                    match future.pop() {
-                        Some((_, frame_id, pipeline_id)) => {
-                            traversal_info.insert(frame_id, pipeline_id);
-                        },
-                        None => return warn!("invalid traversal delta"),
-                    }
+                for entry in self.joint_session_future(top_level_frame_id).take(delta) {
+                    size = size + 1;
+                    table.insert(entry.frame_id, entry);
+                }
+                if size < delta {
+                    return debug!("Traversing forward too much.");
                 }
             },
             TraversalDirection::Back(delta) => {
-                let mut past = self.joint_session_past(top_level_frame_id);
-                for _ in 0..delta {
-                    match past.pop() {
-                        Some((_, frame_id, pipeline_id)) => {
-                            traversal_info.insert(frame_id, pipeline_id);
-                        },
-                        None => return warn!("invalid traversal delta"),
-                    }
+                for entry in self.joint_session_past(top_level_frame_id).take(delta) {
```

Comment:
An extra pointer indirection at runtime and possibly some lost inlining, but the code would be cleaner.